iothub/device/src/Transport/Amqp/AmqpConnectionPool.cs

```diff
@@ -11,108 +11,93 @@ namespace Microsoft.Azure.Devices.Client.Transport.Amqp
             Action<Twin, string, TwinCollection> twinMessageListener, 
-            Func<string, Message, Task> eventListener)
+            Func<string, Message, Task> eventListener,
+            Action onUnitDisconnected)
         {
             if (Logging.IsEnabled) Logging.Enter(this, deviceIdentity, $"{nameof(CreateAmqpUnit)}");
-            if (deviceIdentity.AuthenticationModel != AuthenticationModel.X509 && (deviceIdentity.AmqpTransportSettings?.AmqpConnectionPoolSettings?.Pooling??false))
+            if (deviceIdentity.IsPooling())
             {
-                IAmqpConnectionHolder amqpConnectionHolder;
-                lock (Lock)
+                AmqpConnectionHolder amqpConnectionHolder;
+                lock (_lock)
                 {
-                    ISet<IAmqpConnectionHolder> amqpConnectionHolders = ResolveConnectionGroup(deviceIdentity, true);
-                    if (amqpConnectionHolders.Count < deviceIdentity.AmqpTransportSettings.AmqpConnectionPoolSettings.MaxPoolSize)
-                    {
-                        amqpConnectionHolder = new AmqpConnectionHolder(deviceIdentity);
-                        amqpConnectionHolder.OnConnectionDisconnected += (o, args) => RemoveConnection(amqpConnectionHolders, o as IAmqpConnectionHolder);
-                        amqpConnectionHolders.Add(amqpConnectionHolder);
-                        if (Logging.IsEnabled) Logging.Associate(this, amqpConnectionHolder, "amqpConnectionHolders");
-                    }
-                    else
-                    {
-                        amqpConnectionHolder = GetLeastUsedConnection(amqpConnectionHolders);
-                    }
+                    AmqpConnectionHolder[] amqpConnectionHolders = ResolveConnectionGroup(deviceIdentity);
+                    amqpConnectionHolder = ResolveConnectionByHashing(amqpConnectionHolders, deviceIdentity);
                 }
                 if (Logging.IsEnabled) Logging.Exit(this, deviceIdentity, $"{nameof(CreateAmqpUnit)}");
-                return amqpConnectionHolder.CreateAmqpUnit(deviceIdentity, methodHandler, twinMessageListener, eventListener);
+                return amqpConnectionHolder.CreateAmqpUnit(deviceIdentity, methodHandler, twinMessageListener, eventListener, onUnitDisconnected);
             }
             else
             {
                 if (Logging.IsEnabled) Logging.Exit(this, deviceIdentity, $"{nameof(CreateAmqpUnit)}");
-                return new AmqpConnectionHolder(deviceIdentity)
-                    .CreateAmqpUnit(deviceIdentity, methodHandler, twinMessageListener, eventListener);
+                return new AmqpConnectionHolder(deviceIdentity).CreateAmqpUnit(deviceIdentity, methodHandler, twinMessageListener, eventListener, onUnitDisconnected);
             }
         }
 
-        private void RemoveConnection(ISet<IAmqpConnectionHolder> amqpConnectionHolders, IAmqpConnectionHolder amqpConnectionHolder)
+        public void RemoveAmqpUnit(AmqpUnit amqpUnit)
```

Comment:
a AmqpUnit stands for AMQP resource bundle for a device. Removing an unit will deallocate the resource. Since upper layer will ask unit manager to do the clean up.